### ISSUE TITLE
Creat own plugins and build it with shaka-player

### ISSUE THREAD

**Issue by hiren3897** [google/shaka-player#3493]
**Have you read the [Tutorials](http://shaka-player-demo.appspot.com/docs/api/tutorial-welcome.html)?**
Yes

**Have you read the [FAQ](https://bit.ly/ShakaFAQ) and checked for duplicate open issues?**
YES

**What version of Shaka Player are you using?**
v3.1.1

**Please ask your question**

Hello Team,

I am developing a generic player using shaka-player

where our client has a URL that gives us a long JSON object in response that have many live streams URLs - fetches 1

we do fetch 1

thereafter, we check the end-user DRM support type and pass a parameter on the URL that we get from 1st fetch 1
` eg: URL?format=dashwidevine `  - fetch 2

then we fetch 2 

and then finally we get our manifests with DRM URLs as a response and configure it with the player to be able to load it the player.

for this process, we have to hard code in our application, 
So, to normalize it can we create and build our own plugin in shaka player so that we just have to pass the base URL and it does the other process automatically.

I know this is very bad but I hope I tried my best to explain to you what we need.

Is it possible to do this?

Thank you In advance and please ask a question if there is anything unclear :)

**Comment by joeyparrish** [comment]
This should be doable with a custom URL scheme plugin, and more could be done with a custom manifest parser plugin if needed.  We use these techniques in our integration tests to "stream" simulated content and manifests.

A scheme handler plugin receives a request object and produces an async response object.  These are objects defined by us, not the W3C Fetch API objects of the same name.

With this, you can do something like `player.load('hiren:someid')` where `someid` is anything you can turn into your JSON URL, or even use the JSON URL itself: `player.load('hiren:https://example.com/path/to/json')`.  You should probably also pass the DASH manifest MIME type to the mimeType argument to avoid the need for a HEAD request through your plugin.  See https://shaka-player-demo.appspot.com/docs/api/shaka.Player.html#load

Your plugin gets this URL, parses it, fetches the JSON, and does whatever else it needs to do to get a DASH manifest URL, then returns a final response object containing the DASH manifest.  You can delegate HTTP requests to our NetworkingEngine component to get a response object for free, or you can use Fetch and convert the results. If you build your own, make sure to set the response's `uri` field to the final manifest URI, so that the DASH parser can resolve relative URIs correctly.

This will get you a DASH manifest parsed in a standard way.  If you want to configure the player, for example, with a license server URI from the JSON you fetched, you can do this from your scheme plugin if it captures your player object.

If you want to do more, possibly manipulating, modifying, or supplementing the manifest contents, you can do this with a manifest parser plugin.  Your plugin could delegate the results to our default DASH parser, then body the manifest object from our parser before returning it from yours.

I'm not sure if you need a manifest parser for your purposes, so you should consider doing that only if you absolutely need to change the manifest in a way that can't be replicated through config.

For API details on the scheme plugin, see https://shaka-player-demo.appspot.com/docs/api/shaka.net.NetworkingEngine.html#.registerScheme

If I understand you correctly, the other thing you want to do is to build this plugin into Shaka Player itself.  This is easy to do if you compile it from source:

```sh
python build/build.py --name hiren \
    +@complete +path/to/plugin.js
```

This will create a custom build with all standard components, plus your plugin, in `dist/shaka-player.hiren.js`.  To do this, your plugin will have to conform to Closure Compiler annotations and other compile-time restrictions of the project.

You can run `python build/build.py --help` to see details on what else you can customize at build time.

Does this help?

**Comment by hiren3897** [comment]
Hello @joeyparrish 

Thanks for giving such a great explanation. Yes, it helped me a lot in understanding the concepts.

After reading you correctly I have some doubts about it:

1) to create my own plugin there is no tutorial for it so I can follow [THIS](https://shaka-player-demo.appspot.com/docs/api/shaka.net.DataUriPlugin.html) example?

2) > This will create a custom build with all standard components, plus your plugin, in dist/shaka-player.hiren.js. To do this, your plugin will have to conform to Closure Compiler annotations and other compile-time restrictions of the project.

How we have to conform to Closure Compiler annotations and other compile-time restrictions of the project. once we have our plugin to be built?

3) Once we have the plugin registered we have to link both scripts 
```
for customized UI "shaka-player.ui.js" and for our plugin "shaka-player.hiren.js"?
```
Thank you :)

**Comment by joeyparrish** [comment]
> 1. to create my own plugin there is no tutorial for it so I can follow [THIS](https://shaka-player-demo.appspot.com/docs/api/shaka.net.DataUriPlugin.html) example?

Yes, that's a relatively simple example.  You should also read this tutorial about license server authentication: https://shaka-player-demo.appspot.com/docs/api/tutorial-license-server-auth.html and specifically the part about making new requests through NetworkingEngine.  Search for `authRequest` in that tutorial, and you'll see an example of an async request.  You can use the same technique for delegating to a regular HTTP request within your plugin if you wish.  It would simplify returning a correct `shaka.extern.Response` object.

> 2. > This will create a custom build with all standard components, plus your plugin, in dist/shaka-player.hiren.js. To do this, your plugin will have to conform to Closure Compiler annotations and other compile-time restrictions of the project.
> 
> How we have to conform to Closure Compiler annotations and other compile-time restrictions of the project. once we have our plugin to be built?

If you want to make a custom build of Shaka Player with your plugin _**included**_, **_then_** you need to annotate your code, follow the project's other compiler-enforced rules, and run that `build.py` command to include your plugin in the build of Shaka Player.  But if you want to keep your plugin **_separate_** from Shaka Player and use a standard build of Shaka Player, none of that is required.

> 1. Once we have the plugin registered we have to link both scripts
> 
> ```
> for customized UI "shaka-player.ui.js" and for our plugin "shaka-player.hiren.js"?
> ```

If your plugin is compiled into a custom build of Shaka Player (`shaka-player.hiren.js` in my example), you only have one script that includes **_both_** Shaka and your plugin.

If, however, you want to build your plugin separately, just write plain JavaScript however you like into a separate script, and include that script in your application.  Make sure your plugin script is included after the player script so it can register itself.

Sorry for being unclear before!  Does this help?

**Comment by hiren3897** [comment]
Thank you so much for your time and response :+1: 

I will do the work and let you know if I have any problems further in the development.

Have a good day :100:

**Comment by joeyparrish** [comment]
Happy to help!

**Comment by hiren3897** [comment]
Hello @joeyparrish 

I have a doubt while creating a plugin.

after I pass a JSON URL in `player.load(hiren:http/path./to/json)` it will fetch everything and do things to get the manifest URL and the DRM license. once I have a response object that includes `Manifest URL` and `DRM license URL.

Doubts
- how I have to configure the response so it should play the contents in the player?
-  is it mandatory to return the `shaka.extern.response` object?
- do I have to make a request from the shaka-player network engine to load the manifest and DRM license to play the contents in the player? 
- I can use the same methodology as async request and `shaka.net.HttpPluginUtils.makeResponse` to do so?

**Comment by hiren3897** [comment]
Hello @joeyparrish Can you please clear my doubts about the plugin. when you get your free time. thank you so much.

Have a nice day

**Comment by hiren3897** [comment]
```This will get you a DASH manifest parsed in a standard way. If you want to configure the player, for example, with a license server URI from the JSON you fetched, you can do this from your scheme plugin if it captures your player object.```

I exactly want to do this to configure the player with the DRM license server to be able to play the contents.

Can you tell me how using the scheme plugin I can reach to player object to make this configuration.?

AND

Also can you please clear my earlier doubts :)

Anticipating your response.

thank you :)

**Comment by hiren3897** [comment]
Hello @joeyparrish 
I hope you are well and good :)

I have created the plugin and it returns `shaka.extern.Response` but I get an error in the console saying 

![shakaError](https://user-images.githubusercontent.com/59652404/124764178-f9f7eb80-df34-11eb-9054-5d67d0f23981.png)

My parse method code

```javascript
  static async parse(uri, request, requestType, progressUpdate) {
    try {
      const parts = uri.split(':');
      const path = parts.slice(1).join(':');
      const clientDRM = selectDrm();
      const latestStream = await HexaSchemeJsonPlugin.fetchJsonUrlStreams(path);
      const manifestResponse = await HexaSchemeJsonPlugin.loadManifest(latestStream.streamUrl, clientDRM.drmType);

      const data = shaka.net.HttpFetchPlugin.parse(manifestResponse.uri, request, requestType, progressUpdate);
      const value = await data.promise;
      console.log('value', value);

      /** @type {shaka.extern.Response} */
      const response = {
        uri: value.uri,
        originalUri: value.uri,
        data: value.data,
        headers: value.headers
      };

      console.log(response);
      // TODO drm configurations in player

      // TODO parse response with a standard shaka.networkEngine format
      return shaka.util.AbortableOperation.completed(response);
    } catch (error) {
      return shaka.util.AbortableOperation.failed(error);
    }
  }
```
1) what I have to do to pass the manifest to the player.load() so it is able to play the manifest in the `uri`?

2) can you explain to me how can I configure the DRM license server in the player configuration which I do have in my son response object with manifest URL.

Thank You :)

**Comment by hiren3897** [comment]
Hello @joeyparrish
Can you please comment on this?

Thank you!!

**Comment by TheModMaker** [comment]
You can't use `async` with the `parse` method; it needs to return an `AbortableOperation`, not a `Promise`.  You have two options:

You can use a separate `parseImpl` method if you don't care about it being abortable:

```js
static parse(uri, request, requestType, progressUpdate) {
  return shaka.util.AbortableOperation.notAbortable(this.parseImpl(uri, request, requestType, progressUpdate));
}
static async parseImpl(uri, request, requestType, progressUpdate) {
  // ...
  return response;
}
```

Or you can do it correctly using `chain`.  You have to use old-style Promise chains with `AbortableOperation`:

```js
static parse(uri, request, requestType, progressUpdate) {
  // ...
  return shaka.util.AbortableOperation.notAbortable(HexaSchemeJsonPlugin.fetchJsonUrlStreams(path))
    .chain((latestStream) => {
      return HexaSchemeJsonPlugin.loadManifest(latestStream.streamUrl, clientDRM.drmType);
    })
    .chain((manifestResponse) => {
      return shaka.net.HttpFetchPlugin.parse(manifestResponse.uri, request, requestType, progressUpdate);
    })
    .chain((response) => {
        // Handle response
        return response.
    });
}
```

To handle setting the license server, you can just call `player.configure(...)` in the callback.

**Comment by hiren3897** [comment]
Hello @TheModMaker and @joeyparrish 
Many many thanks for your response 

I tried using the second approach with `chain` but it still gave me the same error with `code:4000`.
![shakaError](https://user-images.githubusercontent.com/59652404/125538153-1f7406a0-2b9d-471b-bbe8-f9020a0dd414.png)
```
CODE
/**
   *
   * @param uri
   * @param request
   * @param requestType
   * @param progressUpdate
   * @return {Promise<shaka.util.AbortableOperation<shaka.extern.Response>|shaka.util.AbortableOperation<any>>}
   */
  static parse(uri, request, requestType, progressUpdate) {
    const parts = uri.split(':');
    const path = parts.slice(1).join(':');
    const clientDRM = selectDrm();

    return shaka.util.AbortableOperation.notAbortable(HexaSchemeJsonPlugin.fetchJsonUrlStreams(path))
      .chain((latestStream) => {
        return HexaSchemeJsonPlugin.loadManifest(latestStream.streamUrl, clientDRM.drmType);
      })
      .chain((manifestResponse) => {
        return shaka.net.HttpFetchPlugin.parse(manifestResponse.uri, request, requestType, progressUpdate);
      })
      .chain((response) => {
        /** @type {shaka.extern.Response} */
        const res = {
          uri: response.uri,
          originalUri: response.uri,
          data: response.data,
          headers: response.headers
        };

        return res;
      });
  }
```
AND
for configuring the DRM license I have to make a callback at the time when I get my final manifest response?
Like example
```
const config = {
  drm: {
    servers: {
      "com.widevine.alpha":`${manifestResponse.drm_primary}`
    }
  }
}

player.configure(config)

```
But here in my plugin js file player doesn't have access to the player object.
Can you please tell me about this?

Thank you so much :)

**Comment by TheModMaker** [comment]
4000 is [UNABLE_TO_GUESS_MANIFEST_TYPE](https://shaka-player-demo.appspot.com/docs/api/shaka.util.Error.html#value:4000).  The problem is we need to know ahead of time which parser to use (DASH or HLS).  We can use a HEAD request to the plugin, but that's complicated.  If you always use one, you can just pass the type to load.

```js
player.load('url', null, 'application/dash+xml');
```

To get the `player` object, you can just capture it when you register the plugin:

```js
const player = ...;
shaka.net.NetworkingEngine.registerScheme('hiren', (uri, request, requestType, progressUpdate) => {
  player.configure(...);
  ...
});
```

Or you can define your plugin in a separate function with an extra argument and forward to that in a captured function:

```js
static parse(player, uri, request, requestType, progressUpdate) {
  ...
}


const player = ...;
shaka.net.NetworkingEngine.registerScheme('hiren', (uri, request, requestType, progressUpdate) => {
  return parse(player, uri, request, requestType, progressUpdate);
});
```

**Comment by hiren3897** [comment]
Hello @TheModMaker 
Thanks for the great guidelines.

when I tried to pass a type to the load it gave me 6012 => `NO_LICENSE_SERVER_GIVEN` which means that I have to configure a license server to load the manifest correctly. Am I right?

And regarding the player object as it should be captured at the time of registering the plugin.

but `const player = ...;` I am confused with this line. what will be replaced or I have to do instead of these `...` dots there to have a player object? 

As I have created a plugin in a separate file so how can I get the player object in that file? 

and I had built my own class that has shaka-player instance init with customized UI. 

```javascript
const player = ...; // what should I do here to have a player object ??
shaka.net.NetworkingEngine.registerScheme('hiren', (uri, request, requestType, progressUpdate) => {
  player.configure(...) // so I can get the drm and configure here!!!
  return HirenPlugin.parse(uri, request, requestType, progressUpdate);
});
```
Thanks, it's a great help for me!! :)))

**Comment by hiren3897** [comment]
Hello @joeyparrish @TheModMaker 
Can you please clear my doubts, thanks.

It would be a great help :)

**Comment by hiren3897** [comment]
Hello, @TheModMaker @joeyparrish can you please help me :) 

thank you

**Comment by theodab** [comment]
Yes, you need to get a player instance in some way. This will either involve creating a player instance on the spot, or getting an existing player instance. It depends on how you are setting the player up.

If you are using [the default setup method for setting up our UI library](https://nightly-dot-shaka-player-demo.appspot.com/docs/api/tutorial-ui.html) (e.g. you have DOM elements with `data-shaka-player-container` and `data-shaka-player` tags in them, as shown in that tutorial), you can get the player instance from the UI library after it has set itself up. That would be like...
```js
// Wait until the UI has loaded.
document.addEventListener('shaka-ui-loaded', () => {
  // Get the video DOM element.
  // TODO: Replace the following line with your own code, if your video does not have the id "video".
  const video = document.getElementById('video');
  // Retrieve the UI instance associated with that video element. 
  const ui = video['ui'];
  // Get the player instance from the UI.
  const player = controls.getControls().getPlayer();
});
```

If you aren't using our UI library, or are using the UI library but setting it up programmatically instead of through tags in the DOM, you will need to create your own player instance. See the section on "programmatic UI setup" [in this tutorial](https://nightly-dot-shaka-player-demo.appspot.com/docs/api/tutorial-ui.html), or [the basic usage tutorial](https://nightly-dot-shaka-player-demo.appspot.com/docs/api/tutorial-basic-usage.html), depending on whether or not you want to use the UI.

**Comment by hiren3897** [comment]
hello @theodab 
yes thanks for the great guidelines it worked for me :tada: 

also thanks to @TheModMaker and @joeyparrish for helping me out.

**Comment by theodab** [comment]
No problem. Good luck with your app, and if you run into any further problems, feel free to ask.

**Comment by hiren3897** [comment]
Yes, Sure Thank you very much :)

**Comment by hiren3897** [comment]
Hello, @theodab I hope you are doing well. can you please help me with #3532 if you have any idea about it :) Thanks
